internal/util/rocksmq/server/rocksmq/rocksmq_retention.go

```diff
@@ -279,23 +282,207 @@ func (ri *retentionInfo) retention() error {
 	for {
 		select {
 		case <-ri.ctx.Done():
-			log.Debug("aaaaaaaaaaaaa")
 			return nil
 		case t := <-ticker.C:
 			timeNow := t.Unix()
 			checkTime := atomic.LoadInt64(&RocksmqRetentionTimeInMinutes) * 60 / 10
 			log.Debug("In ticker: ", zap.Any("ticker", timeNow))
-			ri.lastRetentionTime.Range(func(k, v interface{}) bool {
-				if v.(int64)+checkTime < timeNow {
-					err := ri.expiredCleanUp(k.(string))
+			ri.mutex.RLock()
+			for _, topic := range ri.topics {
+				lastRetentionTsKey := LastRetTsTitle + topic
+				lastRetentionTsVal, err := ri.kv.Load(lastRetentionTsKey)
+				if err != nil {
+					log.Warn("Can't get lastRetentionTs")
+				}
+				lastRetentionTs, _ := strconv.ParseInt(lastRetentionTsVal, 10, 64)
+				if lastRetentionTs+checkTime < timeNow {
+					err := ri.newExpiredCleanUp(topic)
 					if err != nil {
 						log.Warn("Retention expired clean failed", zap.Any("error", err))
 					}
 				}
-				return true
-			})
+			}
+			ri.mutex.RUnlock()
+			// ri.lastRetentionTime.Range(func(k, v interface{}) bool {
+			// 	if v.(int64)+checkTime < timeNow {
+			// 		err := ri.newExpiredCleanUp(k.(string))
+			// 		if err != nil {
+			// 			log.Warn("Retention expired clean failed", zap.Any("error", err))
+			// 		}
+			// 	}
+			// 	return true
+			// })
+		}
+	}
+}
+
+func (ri *retentionInfo) newExpiredCleanUp(topic string) error {
+	log.Debug("Timeticker triggers an expiredCleanUp task for topic: " + topic)
+	ll, ok := topicMu.Load(topic)
+	if !ok {
+		return fmt.Errorf("topic name = %s not exist", topic)
+	}
+	lock, ok := ll.(*sync.Mutex)
+	if !ok {
+		return fmt.Errorf("get mutex failed, topic name = %s", topic)
+	}
+	lock.Lock()
+	defer lock.Unlock()
+
+	var deletedAckedSize int64 = 0
+	var startID UniqueID
+	var endID UniqueID
+	var pageStartID UniqueID = 0
+
+	fixedAckedTsKey, _ := constructKey(AckedTsTitle, topic)
+
+	pageReadOpts := gorocksdb.NewDefaultReadOptions()
+	defer pageReadOpts.Destroy()
+	pageReadOpts.SetPrefixSameAsStart(true)
+	pageIter := ri.kv.DB.NewIterator(pageReadOpts)
+	defer pageIter.Close()
+	pageMsgPrefix, err := constructKey(PageMsgSizeTitle, topic)
+	if err != nil {
+		return err
+	}
+	pageIter.Seek([]byte(pageMsgPrefix))
+	if pageIter.Valid() {
+		pageStartID, err = strconv.ParseInt(string(pageIter.Key().Data())[FixedChannelNameLen+1:], 10, 64)
+		if err != nil {
+			return err
+		}
+
+		for ; pageIter.Valid(); pageIter.Next() {
+			pageID, err := strconv.ParseInt(string(pageIter.Key().Data())[FixedChannelNameLen+1:], 10, 64)
```

Comment:
pageIter.Key().Data() may cause memory leak.
Since you do not call pageIter.Key().Free() ?